test-site/src/components/MapPin.tsx

```diff
@@ -1,9 +1,48 @@
-import { PinComponent } from '@yext/search-ui-react';
+import { useEffect, useRef, useState } from 'react';
+import ReactDOM from 'react-dom';
+import { Map, Popup, LngLatLike } from 'mapbox-gl';
+import { PinComponent, Coordinate } from '@yext/search-ui-react';
+import { Result } from '@yext/search-headless-react';
 import { Location } from '../pages/LocationsPage';
+import { MapPopup } from './MapPopup';
+
+interface MapPinProps {
+  result: Result<Location>
+  index: number
+  mapbox: Map
+}
+
+const transformToMapboxCoord = (coordinate: Coordinate): LngLatLike => ({
+  lng: coordinate.longitude,
+  lat: coordinate.latitude,
+});
+
+export const MapPin: PinComponent<Location> = (props: MapPinProps) => {
+  const popupRef = useRef(new Popup({ offset: 15 }));
+  const initialRender = useRef(true);
+  const [active, setActive] = useState(false);
+
+  useEffect(() => {
+    if (!initialRender.current && active) {
+      const popup = document.createElement('div');
+      ReactDOM.render(<MapPopup entity={props.result} />, popup);
+      popupRef.current
+        .setLngLat(coordinate)
+        .setDOMContent(popup)
+        .addTo(props.mapbox);
+    }
+    initialRender.current = false;
+  }, [active]);
+
+  const yextCoordinate: Location["yextDisplayCoordinate"] = props.result.rawData.yextDisplayCoordinate;
+  if (!yextCoordinate) return <></>;
+  const coordinate = transformToMapboxCoord(yextCoordinate);
 
-export const MapPin: PinComponent<Location> = () => {
   return (
-    <div className='relative z-10'>
+    <button
+      className='relative z-10'
+      onClick={() => setActive(!active)}
```

Comment:
I might be misreading but I think here `active` doesn't actually determine the active state? I checked out this branch out of curiosity and it looks like that after the initial render of the popup, if you click out of it to close it, you need to click twice on the button to reopen it? I assume this is because 
1. when you click outside of the Popup mapbox automatically calls Popup.remove() and
2. the first click on the button changes active to false, which doesn't do anything, and only on the second click does it switch back to true which then re-adds the popup

Reply:
renamed to something clearer. I can only get the mapboxgl.Map to update in useEffect [according to their docs](https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/), and my experience trying alternatives. If you have any other ideas I'd be happy to try, this feels a little weird.

Reply:
Reverted back to `active` and added an `on close` event to the popup following docs here. Like this much better https://docs.mapbox.com/mapbox-gl-js/api/markers/#popup.event:close